content/blog/2018-02-07-update-on-async-rendering.md

```diff
@@ -0,0 +1,173 @@
+
+People often assume that `componentWillMount` and `componentWillUnmount` are always paired, but that is not guaranteed. Only once `componentDidMount` has been called does React guarantee that `componentWillUnmount` will later be called for clean up.
+
+For this reason, the recommended way to add listeners/subscriptions is to use the `componentDidMount` lifecycle:
+`embed:update-on-async-rendering/adding-event-listeners-after.js`
+
+> Note
+>
+> Although the pattern above is slightly more verbose, it has an added benefit of deferring the subscription creation until after the component has rendered, reducing the amount of time in the critical render path. In the near future, React may include more tools to reduce code complexity for data fetching cases like this.
+
+> Note
+>
+> Sometimes it is important to update subscriptions in response to property changes. If you're using a library like Redux or MobX, the library's container component should handle this for you. If you're authoring such a library, we suggest using a technique like [the one shown here](https://gist.github.com/bvaughn/d569177d70b50b58bff69c3c4a5353f3).
+
+### Updating `state` based on `props`
+
+Here is an example of a component that uses the legacy `componentWillReceiveProps` lifecycle to update `state` based on new `props` values:
+`embed:update-on-async-rendering/updating-state-from-props-before.js`
+
+Although the above code is not problematic in itself, the `componentWillReceiveProps` lifecycle is often mis-used in ways that _do_ present problems. Because of this, the method has been deprecated.
+
+As of version 16.3, the recommended way to update `state` in response to `props` changes is with the new `static getDerivedStateFromProps` lifecycle. (That lifecycle is called when a component is created and each time it receives new props.):
+`embed:update-on-async-rendering/updating-state-from-props-after.js`
+
+> Note
+>
+> The [`react-lifecycles-compat`](https://github.com/reactjs/react-lifecycles-compat) polyfill enables this new lifecycle to be used with older versions of React as well. [Learn more about how to use it below.](http://localhost:8000/blog/2018/02/07/update-on-async-rendering.html#open-source-project-maintainers)
+
+### Invoking external callbacks
+
+Here is an example of a component that calls an external function when its internal state changes:
+`embed:update-on-async-rendering/invoking-external-callbacks-before.js`
+
+Sometimes people use `componentWillUpdate` out of a misplaced fear that by the time `componentDidUpdate` fires, it is "too late" to update the state of other components. This is not the case. React ensures that any `setState` calls that happen during `componentDidMount` and `componentDidUpdate` are flushed before the user sees the updated UI. In general, it is better to avoid cascading updates like this, but in some cases they are necessary (for example, if you need to position a tooltip after measuring the rendered DOM element).
```

Comment:
Are `setState` calls being flushed to the UI before the user sees it a new feature? Assumed in current React that `didMount` meant it's in the DOM and it's visible, same with `didUpdate`

Reply:
No. This has always been the case.

Libraries like react-virtualized depend on this behavior, for example, by measuring DOM elements in `componentDidMount`/`componentDidUpdate` and potentially re-rendering. The additional render is processed synchronously, so the user doesn't see the transient UI state.

Reply:
Jeez, that is news to me. I've got to write about this now. 100% my assumption has always been "mounted/update" = "visible to user".
Glad that's cleared up. I assume I'm not alone in this assumption.

Reply:
Well, it is "visible" in that it's in the DOM, but the sync update is processed immediately so the user doesn't actually have time to _see_ it.